JamGame/Assets/Scripts/Employee/Controller/ControllerImpl.cs

```diff
@@ -73,18 +72,49 @@ private void Update()
                     if (agent.remainingDistance < 0.01f)
                     {
                         state = State.Idle;
-                        OnFinishedMoving?.Invoke();
+                        transform.position = currentDestination + (agent.baseOffset * Vector3.up);
+                        OnReachedNeedProvider?.Invoke();
                     }
                     break;
             }
         }
 
+        public void SetDestination(Vector3 target_position)
+        {
+            currentDestination = target_position;
+            _ = agent.SetDestination(currentDestination);
+            state = State.BuildingPath;
+        }
+
+        public void SetNavigationMode(NavigationMode mode)
+        {
+            if (navigationMode == mode)
+            {
+                return;
+            }
+
+            navigationMode = mode;
+            agent.enabled ^= true;
+
+            if (mode == NavigationMode.FreeMove)
+            {
+                state = State.Idle;
+            }
+            else
+            {
+                state = State.BuildingPath;
+            }
+        }
+
         private void CorrectMovement()
         {
-            agent.speed =
+            float max_speed =
                 (1.0f - personalSpace.GetCrowdMetrics())
                 * maxVelocity
                 * maxVelocityMultiplierByEffects;
+            max_speed = Mathf.Clamp(max_speed, 0.0f, float.MaxValue);
+
+            agent.speed = max_speed;
 
             Vector3 steering = personalSpace.GetPreferredSteeringNormalized();
             if (steering.sqrMagnitude > 0.0001)
```

Comment:
What is this constant `0.0001`?

Reply:
Just some small enough number as I'm not shure that float.EPSILON will work correctly in this context